### PR TITLE
Generate data.json and UI fields automatically; update config

### PR DESCRIPTION
**Pull request by axelchalon**
https://github.com/paritytech/parity-config-generator/issues/34

Let me know what you think

Parses https://github.com/paritytech/parity/blob/master/parity/cli/mod.rs and augments it  with `data.extra.json` to generate `data.compiled.json`. Now we're sure that default values &c. are up-to-date (so long as we run the script and make sure that data.extra.json is up-to-date). UI fields are now automatically generated. This way the items are automatically added or removed from the UI if there are new/removed config options.

(I kept the help messages of master in data.extra.json because I assumed they were more user-friendly than the CLI help messages, but I don't know if they're all still up-to-date.)

### CODE REVIEW

File: scripts/generate-config-data.js [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+    [
+      /^\|c: &Config\| c.([a-zA-Z_]+).as_ref\(\)\?\.([a-zA-Z_]+)(?:\.clone\(\))?$/,
+      'same'
+    ],
+    [
+      /^\|c: &Config\| c.([a-zA-Z_]+).as_ref\(\)\?\.([a-zA-Z_]+)(?:\.clone\(\))?\.map\(\|([a-zA-Z_]+)\| !\3\)(\.clone\(\))?$/,
+      'opposite'
+    ],
+    [
+      /^\|c: &Config\| c.([a-zA-Z_]+).as_ref\(\)\?\.([a-zA-Z_]+)(?:\.clone\(\)|\.as_ref\(\))?\.map\(\|vec\| vec.join\(","\)\)?$/,
```
Comment (tomusdrw) [inline scripts/generate-config-data.js]: Perhaps we could ignore everything between `&Config` and `\.join\(`? this should make the regex simpler and more future-proof.

File: scripts/generate-config-data.js [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+          dataProp.default = cliConfigProp.defaultValue;
+        } else if (cliConfigProp.configIs === 'array_of') {
+          dataProp.description = cliConfigProp.help;
+          if (cliConfigProp.defaultValue !== null) {
+            dataProp.default = cliConfigProp.defaultValue.split(',');
+          }
+        } else if (cliConfigProp.configIs === 'opposite') {
+          dataProp.default = !cliConfigProp.defaultValue;
+        } else {
+          console.error("Error: Unreachable: configIs can be 'same', 'array_of', 'opposite', 'none', 'unknown'; is none of the first three; last two are pruned in makeCliConfigTree; qed.");
```
Comment (tomusdrw) [inline scripts/generate-config-data.js]: Would turn it into hard error (`throw new Error()`) and display `cliConfigProp.configIs` in the error as well.

File: scripts/generate-config-data.js [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+    if (!id.includes('.')) {
+      dataAugmentedOrdered[id] = dataAugmented[id];
+    } else {
+      const [section, prop] = id.split('.');
+      dataAugmentedOrdered[section] = dataAugmentedOrdered[section] || {};
+      dataAugmentedOrdered[section][prop] = dataAugmented[section][prop];
+    }
+  });
+
+  for (const section in dataAugmented) {
```
Comment (tomusdrw) [inline scripts/generate-config-data.js]: Not a fan of such loops, but can live with this. Would prefer:
```
Object.keys(dataAugmented)
  .filter(section => !overwritten.includes(section))
  .forEach/map()
```

File: scripts/generate-config-data.js [hunk 4]
```diff
@@ -0,0 +1,245 @@
+const https = require('https');
+const fs = require('fs');
+const path = require('path');
+
+function fetchSource () {
+  return new Promise(resolve => {
+    https.get('https://raw.githubusercontent.com/paritytech/parity/master/parity/cli/mod.rs', res => {
+      res.setEncoding('utf8');
+      let body = '';
+      res.on('data', data => { body += data; });
```
Comment (tomusdrw) [inline scripts/generate-config-data.js]: Should handle error.

Comment (tomusdrw) [inline scripts/generate-config-data.js]: Could use https://github.com/rvagg/bl and `require('util').promisify` to simplify processing.

File: scripts/generate-config-data.js [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+    const defaultValue = parseDefaultValue(defaultValueString);
+    const help = helpString.replace(/\\"/g, '"');
+    const {configSection, configProp, configIs} = parseConfigCallback(configCallbackString);
+    return {defaultValue, help, configSection, configProp, configIs, ...rest};
+  });
+
+  return parsedCliOptions;
+}
+
+function* execAll (regex, str) {
```
Comment (tomusdrw) [inline scripts/generate-config-data.js:50]: Neat! :+1:

File: scripts/generate-config-data.js [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+  switch (defaultValue) {
+    case 'None':
+      return null;
+    case 'Vec::new()':
+      return [];
+    default:
+      try {
+        return JSON.parse(defaultValue);
+      } catch (e) {
+        console.warn('Warning: Failed to parse default value', defaultValue);
```
Comment (tomusdrw) [inline scripts/generate-config-data.js:68]: Should it return something different than `undefined`?

File: scripts/generate-config-data.js [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+    }
+  }
+
+  console.warn('Warning: Failed to recognize callback', configCallback);
+  return {configIs: 'unknown'};
+}
+
+function makeCliConfigTree (parsedCliOptions) {
+  const configCliOptions = parsedCliOptions.filter(({configIs}) =>
+        configIs !== 'none' && configIs !== 'unknown'
```
Comment (tomusdrw) [inline scripts/generate-config-data.js]: Perhaps use different indentation:
```
const configCliOptions = parsedCliOptions
  .filter(({configIs}) => configIs !== 'none' && configIs !== 'unknown')

File: scripts/generate-config-data.js [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+      data[section.name][prop.name].name = prop.name.replace(/_/g, ' ').replace(/^./, a => a.toUpperCase());
+      data[section.name][prop.name].type = prop.type;
+
+      // If the config prop is linked to a CLI command
+      if (section.name in cliConfig &&
+          prop.name in cliConfig[section.name]) {
+        const cliConfigProp = cliConfig[section.name][prop.name];
+        const dataProp = data[section.name][prop.name];
+
+        if (cliConfigProp.configIs === 'same') {
```
Comment (tomusdrw) [inline scripts/generate-config-data.js]: Would move `same`, `array_of`, etc to constants.

File: scripts/generate-config-data.js [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+
+  // Augment with data.extra.json
+
+  const extra = JSON.parse(fs.readFileSync(path.resolve(__dirname, '../src/data.extra.json'), 'UTF-8'));
+  const dataAugmented = augment(data, extra);
+
+  // Write to file
+
+  fs.writeFileSync(path.resolve(__dirname, '../src/data.compiled.json'), JSON.stringify(dataAugmented, null, 2));
+})();
```
Comment (tomusdrw) [inline scripts/generate-config-data.js]: Should handle errors from the returned promise (by `console.error` and `process.exit(1)`)

File: src/components/Editor.js [hunk 10]
```diff
@@ -23,237 +17,118 @@ class Editor extends Component {
...
     return (
       <div>
         { this.select('__internal', 'platform') }
         { this.select('__internal', 'configMode') }
-        <div style={isSimple ? styles.visible : styles.hidden}>
-          { this.renderSimple(settings, platform, base, isOffline) }
-        </div>
-        <div style={!isSimple ? styles.visible : styles.hidden}>
-          { this.renderConfig(settings, platform, base, isOffline) }
+        <div>
```
Comment (tomusdrw) [inline src/components/Editor.js:50]: Where did the simplified version go :( IMHO we should auto-generate the full one, but keep the simple one hardcoded.

**Review by tomusdrw — COMMENTED** [review]
Looks good, couple of grumbles.

File: scripts/generate-config-data.js (hunk 1 above)
Comment (axelchalon) [reply scripts/generate-config-data.js]: hmm but we have capturing parentheses between `&Config` and `\.join\(`, e.g. ` |c: &Config| c.(((ui))).as_ref()?.(((hosts))).as_ref().map(|vec| vec.join(",")),`; this is how we know the config section & prop associated with this CLI option (and thus the config default value based on the CLI default value)

File: scripts/generate-config-data.js (hunk 4 above)
Comment (axelchalon) [reply scripts/generate-config-data.js]: cool! didn't know about `bl`

However I ended up with the following structure and I couldn't see how I could fit `promisify` in there to further simplify:

```javascript
function fetchSource () {
  return new Promise((resolve, reject) => {
    https.get('https://raw.githubusercontent.com/paritytech/parity/master/parity/cli/mod.rs', res => {
      if (res.statusCode !== 200) {
        reject(new Error(`Failed to load source code, status code: ${res.statusCode}`));
      }

      res.pipe(bl(function (err, data) {
        if (err) {
          reject(new Error(err));
        }

        resolve(data.toString());
      }))
    });
  });
}
```

File: scripts/generate-config-data.js (hunk 6 above)
Comment (axelchalon) [reply scripts/generate-config-data.js:68]: I think it's fine; also lets us distinguish between `null` default values and unrecognized (undefined) return values (although we could admittedly remove the `defaultValue` key altogether in the latter case).

The idea anyway is to provide a default value in data.extra.json for the config items that throw this warning.
I guess we could issue an error if a config item with unrecognized default value doesn't have a default value set in data.extra.json. I'll integrate this to the next commit.

File: src/components/Editor.js (hunk 10 above)
Comment (axelchalon) [reply src/components/Editor.js:50]: :D 

We still have a simple version (only shows items/sections with `.simple === true` in data.extra.json). The only differences with `master` are the `Parity` & `Footprint` sections coming first and not being collapsible, and the custom `Servers` section on `master` that lists the port/interface settings from the different sections.

- `Parity` & `Footprint` first and not collapsible: we could A) reorder the sections in data.extra.json to make `Parity` and `Footprint` appear first regardless of the configMode, and add `simple_collapsible: false` to those sections in data.extra.json; or B) tweak `renderConfig` to render those sections first in simple mode and without the collapsible markup.
- custom `Servers` section: we could A) add `simple_section: 'Servers'` to the concerned props in `data.extra.json` and tweak `renderConfig` accordingly (order would be: non-collapsible sections; custom sections; sections and the order would be the same as `master`)

What do you think? We can go another way if you prefer

**Pull request merged**